### PR TITLE
[REEF-1681] Allow REEF Driver to run in unmanaged AM mode

### PR DESCRIPTION
**Pull request by motus**
* Introduce `DriverIsUnmanaged` config parameter
  * Set `ApplicationSubmissionContext.setUnmanagedAM()` when submitting application request to YARN
  * Minor refactoring for readability and better logging
  * Additional javadoc comments

JIRA: [REEF-1681](https://issues.apache.org/jira/browse/REEF-1681)

### CODE REVIEW

**Comment by motus** [comment]
@markusweimer I've rolled back most of the style changes + moved some boilerplate code to utils. Please take a look!

File: lang/java/reef-common/src/main/java/org/apache/reef/util/TangUtil.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+import org.apache.reef.tang.Configuration;
+import org.apache.reef.tang.Tang;
+import org.apache.reef.tang.annotations.Name;
+import org.apache.reef.tang.exceptions.InjectionException;
+
+/**
+ * A collection of helper functions to extract values from Tang configuration.
+ */
+public final class TangUtil {
```
Comment (tcNickolas) [inline lang/java/reef-common/src/main/java/org/apache/reef/util/TangUtil.java]: Would this class be more fitting in reef-tang project?

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java [hunk 2]
```diff
@@ -117,38 +117,38 @@ public String getStringApplicationId() {
...
+    this.applicationSubmissionContext.setResource(Resource.newInstance(getMemory(megabytes), 1));
     return this;
   }
 
   /**
    * Add a file to be localized on the driver.
-   * @param resourceName
-   * @param resource
-   * @return
+   * @param resourceName String name of the resource.
```
Comment (tcNickolas) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java]: `String` should be lowercase probably

File: lang/java/reef-common/src/main/java/org/apache/reef/driver/parameters/DriverIsUnmanaged.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+import org.apache.reef.tang.annotations.Name;
+import org.apache.reef.tang.annotations.NamedParameter;
+
+/**
+ * If true, run Driver in unmanaged mode.
+ * (i.e. co-located with the client and not managed by the RM).
+ * This setting is false by default.
+ */
+@NamedParameter(doc = "Run driver in unmanaged mode", default_value = "false")
```
Comment (tcNickolas) [inline lang/java/reef-common/src/main/java/org/apache/reef/driver/parameters/DriverIsUnmanaged.java]: If default value of unmanaged is false, should we call the parameter (and all matching variables) managed? The fewer negations there are, the easier it is to parse things for the reader

File: lang/java/reef-common/src/main/java/org/apache/reef/util/TangUtil.java (hunk 1 above)
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/util/TangUtil.java]: Does this class have to be `public`? At the very least mark it `@Private` to not add it to the supported REEF surface.

File: lang/java/reef-common/src/main/java/org/apache/reef/util/TangUtil.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+  /**
+   * Extract a parameter's value from Tang configuration.
+   * @param config Tang configuration.
+   * @param clazz A parameter to extract from the configuration.
+   * @param defaultValue default value to use if the parameter is missing.
+   * @param <T> Type of the parameter - inferred from the parameter's class.
+   * @return Value of the configuration parameter, or the default value
+   * if it is not present in the Tang configuration.
+   */
+  public static <T> T getNamedInstance(
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/util/TangUtil.java]: Methods like this make me nervous, as they essentially negate the use of Tang and turn a `Configuration` instance into a mere hashmap. What is the reason for needing it? Maybe we can work on it at the callsite.

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnJobSubmissionHandler.java [hunk 5]
```diff
@@ -90,19 +93,23 @@ public void close() {
   @Override
   public void onNext(final JobSubmissionEvent jobSubmissionEvent) {
 
-    LOG.log(Level.FINEST, "Submitting job with ID [{0}]", jobSubmissionEvent.getIdentifier());
+    LOG.log(Level.FINER, "Submitting {0} job with ID {1}",
+        new Object[] {this.isDriverUnmanaged ? "UNMANAGED" : "regular", jobSubmissionEvent.getIdentifier()});
```
Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnJobSubmissionHandler.java]: Maybe harmonize this to `unmanaged` and `managed`? `regular` tends to loose meaning over time.

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnJobSubmissionHandler.java [hunk 6]
```diff
@@ -90,19 +93,23 @@ public void close() {
...
 
     try (final YarnSubmissionHelper submissionHelper =
              new YarnSubmissionHelper(this.yarnConfiguration, this.fileNames, this.classpath, this.tokenProvider)) {
 
-      LOG.log(Level.FINE, "Assembling submission JAR for the Driver.");
-      final Optional<String> userBoundJobSubmissionDirectory =
-          getUserBoundJobSubmissionDirectory(jobSubmissionEvent.getConfiguration());
+      LOG.log(Level.FINEST, "Assembling submission JAR for the Driver.");
```
Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnJobSubmissionHandler.java]: This change looks too big on GitHub. Shouldn't we just see `FINE` replaced with `FINEST`?

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnJobSubmissionHandler.java [hunk 7]
```diff
@@ -90,19 +93,23 @@ public void close() {
...
+
+      final Optional<String> userBoundJobSubmissionDirectory = getUserBoundJobSubmissionDirectory(jobSubmissionEvent);
+
       final JobFolder jobFolderOnDfs = userBoundJobSubmissionDirectory.isPresent()
           ? this.uploader.createJobFolder(userBoundJobSubmissionDirectory.get())
           : this.uploader.createJobFolder(submissionHelper.getApplicationId());
+
```
Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnJobSubmissionHandler.java]: What changed here?

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnJobSubmissionHandler.java [hunk 8]
```diff
@@ -90,19 +93,23 @@ public void close() {
...
       final Configuration driverConfiguration = makeDriverConfiguration(jobSubmissionEvent, jobFolderOnDfs.getPath());
       final File jobSubmissionFile = this.jobJarMaker.createJobSubmissionJAR(jobSubmissionEvent, driverConfiguration);
+
```
Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnJobSubmissionHandler.java]: What changed here?

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnJobSubmissionHandler.java [hunk 9]
```diff
@@ -112,11 +119,14 @@ public void onNext(final JobSubmissionEvent jobSubmissionEvent) {
...
           .setPriority(getPriority(jobSubmissionEvent))
           .setQueue(getQueue(jobSubmissionEvent))
+          .setUnmanagedAM(this.isDriverUnmanaged)
           .setPreserveEvaluators(getPreserveEvaluators(jobSubmissionEvent))
           .setMaxApplicationAttempts(getMaxApplicationSubmissions(jobSubmissionEvent))
           .submit();
 
-      LOG.log(Level.FINEST, "Submitted job with ID [{0}]", jobSubmissionEvent.getIdentifier());
+      LOG.log(Level.FINE, "Submitted {0} job with ID {1}",
+          new Object[] {this.isDriverUnmanaged ? "UNMANAGED" : "regular", jobSubmissionEvent.getIdentifier()});
```
Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnJobSubmissionHandler.java]: Same as above. Don't use `regular`

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java [hunk 10]
```diff
@@ -41,37 +41,37 @@
...
-  private final YarnClientApplication yarnClientApplication;
   private final GetNewApplicationResponse applicationResponse;
   private final ApplicationSubmissionContext applicationSubmissionContext;
   private final ApplicationId applicationId;
   private final Map<String, LocalResource> resources = new HashMap<>();
-  private final REEFFileNames fileNames;
   private final ClasspathProvider classpath;
   private final SecurityTokenProvider tokenProvider;
   private final List<String> commandPrefixList;
+
```
Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java]: What changed here?

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java [hunk 11]
```diff
@@ -41,37 +41,37 @@
...
                               final ClasspathProvider classpath,
                               final SecurityTokenProvider tokenProvider,
                               final List<String> commandPrefixList) throws IOException, YarnException {
-    this.fileNames = fileNames;
+
     this.classpath = classpath;
 
     this.driverStdoutFilePath =
-        ApplicationConstants.LOG_DIR_EXPANSION_VAR + "/" + this.fileNames.getDriverStdoutFileName();
+        ApplicationConstants.LOG_DIR_EXPANSION_VAR + "/" + fileNames.getDriverStdoutFileName();
```
Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java]: Why this change? I thought we used `this.` notation?

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java [hunk 12]
```diff
@@ -41,37 +41,37 @@
...
 
     this.driverStderrFilePath =
-        ApplicationConstants.LOG_DIR_EXPANSION_VAR + "/" + this.fileNames.getDriverStderrFileName();
+        ApplicationConstants.LOG_DIR_EXPANSION_VAR + "/" + fileNames.getDriverStderrFileName();
```
Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java]: Why this change? I thought we used `this.` notation?

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java [hunk 13]
```diff
@@ -41,37 +41,37 @@
...
-  private final REEFFileNames fileNames;
   private final ClasspathProvider classpath;
   private final SecurityTokenProvider tokenProvider;
   private final List<String> commandPrefixList;
+
   private String driverStdoutFilePath;
   private String driverStderrFilePath;
-  private Class launcherClazz;
   private List<String> configurationFilePaths;
+  private Class launcherClass = REEFLauncher.class;
```
Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java]: Can this be `final`?

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java [hunk 14]
```diff
@@ -80,21 +80,21 @@ public YarnSubmissionHelper(final YarnConfiguration yarnConfiguration,
...
-    this.configurationFilePaths = Collections.singletonList(this.fileNames.getDriverConfigurationPath());
+    this.configurationFilePaths = Collections.singletonList(fileNames.getDriverConfigurationPath());
     LOG.log(Level.INFO, "YARN Application ID: {0}", this.applicationId);
   }
 
   public YarnSubmissionHelper(final YarnConfiguration yarnConfiguration,
                               final REEFFileNames fileNames,
                               final ClasspathProvider classpath,
                               final SecurityTokenProvider tokenProvider) throws IOException, YarnException {
+
```
Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java]: What changed here?

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java [hunk 15]
```diff
@@ -171,48 +171,60 @@ public YarnSubmissionHelper setPreserveEvaluators(final boolean preserveEvaluato
...
   }
 
   /**
+   * Specify whether the application manager should be run in unmanaged mode.
+   * If true, the Driver will run locally but still use YARN to allocate Evaluators.
+   * By default this is false, i.e. REEF Driver runs in its own container managed by YARN.
+   * @param isUnmanaged boolean flag whether to run Driver in unmanaged mode.
+   * @return reference to self for chain calls.
+   */
+  public YarnSubmissionHelper setUnmanagedAM(final boolean isUnmanaged) {
```
Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java]: Do we need the parameter? As in, can't `setUnmanagedAM` just set the Driver to be unmanaged?

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java [hunk 16]
```diff
@@ -285,19 +299,19 @@ public void submit() throws IOException, YarnException {
...
-          "Requested {0}MB of memory for the driver. " +
-              "The max on this YARN installation is {1}. " +
-              "Using {1} as the memory for the driver.",
-          new Object[]{requestedMemory, maxMemory});
+      LOG.log(Level.WARNING, "Requested {0}MB of memory for the driver. " +
+          "The max on this YARN installation is {1}. Using {1} as the memory for the driver.",
+          new Object[] {requestedMemory, maxMemory});
       amMemory = maxMemory;
     }
+
```
Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java]: What changed here?

**Comment by motus** [comment]
OK let's close it for now. Thanks for reviewing, though - I will incorporate the changes into another more comprehensive PR later.

**Comment by tcNickolas** [comment]
I don't like the concept of closing pull request with 20+ comments and incorporating it into another one. It will be a pain for the reviewer to recover this pull request and check that the comments have actually been addressed. Is it possible to create a sub-task of REEF-1681 which focuses on introducing new parameter, and continue this PR to close it?

**Comment by motus** [comment]
What I meant is that PR does not change that much, but blocks more important functionality that I am working on right now. I would much prefer to not sweat over details of this PR now, but if you insist, I will reopen it and reduce the change to the minimum. (a new commit will follow shortly)

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java (hunk 13 above)
Comment (motus) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java]: it cannot. we may override it later in `.setLauncherClass()` method

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java (hunk 11 above)
Comment (motus) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java]: `fileNames` is not a member variable anymore (it does not have to be)

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java (hunk 12 above)
Comment (motus) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java]: same here. I've got rid of `fileNames` as a member variable - now it's a constructor parameter only

File: lang/java/reef-common/src/main/java/org/apache/reef/driver/parameters/DriverIsUnmanaged.java (hunk 3 above)
Comment (motus) [inline lang/java/reef-common/src/main/java/org/apache/reef/driver/parameters/DriverIsUnmanaged.java]: I thought about it, and I think "unmanaged" would be more appropriate, because "Unmanaged AM" is a special mode for YARN applications. OTOH, there is no designation for "managed" mode in YARN, and it would be very confusing to introduce the term

**Comment by motus** [comment]
@markusweimer @tcNickolas Here goes the new version!

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java (hunk 10 above)
Comment (motus) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java]: er, an empty line that separates final and non-final member variables?

File: lang/java/reef-common/src/main/java/org/apache/reef/driver/parameters/DriverIsUnmanaged.java (hunk 3 above)
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/driver/parameters/DriverIsUnmanaged.java]: Let's keep it as `unmanaged` then, as it is the only proper noun that exists.

**Comment by tcNickolas** [comment]
This version of the change looks much smaller than the previous one :-) Did you decide to drop `TangUtils` and extra logging you had previously?

Other than that, LGTM. I can merge

**Comment by motus** [comment]
@tcNickolas yes, I agree with @markusweimer that `TangUtils` is a hack in any form. Let's keep that functionality hidden so there is no temptation to reuse it elsewhere. Also, there was not much extra logging anyway, and it is likely I will do more logging in other modules in the subsequent PRs.

**Comment by tcNickolas** [comment]
I've merged this
Sorry, I forgot to put in PR # in the commit message - could you please close it manually?

**Comment by motus** [comment]
Merged into master
